src/tools/crawler/crawl.rs

```diff
@@ -0,0 +1,121 @@
+//! Contains structs and methods to crawl ripple network according to instruction at https://xrpl.org/peer-crawler.html
+
+use std::{
+    net::{IpAddr, SocketAddr},
+    num::ParseIntError,
+    time::Duration,
+};
+
+use reqwest::{Client, StatusCode};
+use serde::Deserialize;
+use thiserror::Error;
+use tokio::time::Instant;
+
+const CRAWLER_DEFAULT_PORT: u16 = 51235;
+const CRAWLER_TIMEOUT: Duration = Duration::from_secs(10);
+
+#[allow(dead_code)]
+#[derive(Debug, Deserialize)]
+pub(super) struct Peer {
+    pub(super) complete_ledgers: Option<String>,
+    pub(super) complete_shards: Option<String>,
+    pub(super) ip: Option<String>,
+    pub(super) port: Option<Port>,
+    pub(super) public_key: String,
+    #[serde(rename = "type")]
+    pub(super) connection_type: String,
+    pub(super) uptime: u32,
+    pub(super) version: String,
+}
+
+impl Peer {
+    /// Returns port number for the peer. If no port was present in the response then returns the default port.
+    /// Returns [ParseIntError] if the response contained invalid value.
+    pub(super) fn port(&self) -> Result<u16, ParseIntError> {
+        self.port
+            .as_ref()
+            .map(|p| match p {
+                Port::Number(n) => Ok(*n),
+                Port::String(s) => s.parse(),
+            })
+            .unwrap_or(Ok(CRAWLER_DEFAULT_PORT))
```

Comment:
this means an error will actually never be returned, so the return value either should not be a `Result` or there should be a `?` after the call to `parse`

Reply:
I've added a test to show when error is returned (bottom of the file).

Reply:
ah, I missed the `as_ref`; ok, so it's an `Option<T>` turned into an `Option<Result<T>>`, unwrapped from the initial option at the end

Reply:
Yes. Port can omitted in the response in which case I'm using the default. If there is a string - it tries to parse it.